README
Sentry::Raven version 1.14

Sentry::Raven is a module for sending events to a sentry service.

SYNOPSIS

  use Sentry::Raven;
  my $raven = Sentry::Raven->new(sentry_dsn => 'https://<publickey>:<secretkey>@sentry.io/<projectid>' );
  $raven->capture_message('The sky is falling');

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

Perl 5.10 or later.

COPYRIGHT AND LICENSE

Copyright (C) 2020 by Matt Harrington

The full text of this license can be found in the LICENSE file included with this module.


Changes
1.14 2020-10-11
 - Do not fail when attempting to read unreadable files (thanks Gianni
   Ceccarelli)

1.13 2020-06-08
 - Fixing precidence of sentry dsn parameters (thanks Nowaker)
 - Displays source code in sentry event (thanks dariuszpiwowarski)

1.12 2020-02-02
 - Sending messages fall back to interface contexts (thanks jrubinator)

1.11 2019-06-29
 - Updating documentation

1.10 2017-06-12
 - Improving user context (thanks Steve Wooster)

1.09 2017-02-19
 - Increase MAX_EXCEPTION_VALUE to 4096 (thanks Michael Wiencek)
 - Using Time::Piece (thanks Ryutaro Mizokami)

1.08 2016-08-05
 - Fixing perl 5.8 compatibility

1.07 2016-08-04
 - Adding environment context parameter (thanks William Travis Holton)

1.06 2016-07-06
 - Adding release context parameter (thanks Slobodan Mišković)

1.05 2016-05-25
 - Fixing stack frame context (thanks Steve Wooster)

1.04 2016-05-10
 - Adding event fingerprinting (thanks Steve Wooster)

1.03 2016-04-15
 - Regain ability to set encoding (thanks Stuart Skelton)

1.02 2016-01-12
 - Move project to getsentry github repo
 - Change parameter reporting (thanks dakkar)

1.01 2015-08-20
 - Clean up encoding error in Sentry

1.00 2015-07-13
 - Major release
 - using stacktrace abs_path and filename correctly

0.09 2015-01-20
 - Reuse TCP connections when possible
 - Use LWP's built-in gzip compression

0.08 2014-09-22
 - Trimming fields to avoid 413 Request Entity Too Large response
 - Fixing Http interface documentation for data field

0.07 2014-08-11
 - Use IO::Compress::Gzip to compress http posts

0.06 2014-07-14
 - Report post errors in capture_errors
 - Added travis config

0.05 2014-06-12
 - Windows test fixes

0.04 2014-06-11
 - Event processor support
 - Stacktrace context now accepts Devel::StackTrace objects
 - Improving error handler behavior and documentation
 - Better context usability for extra and tags

0.03 2014-05-29
 - Context decorators
 - Post users (sentry interface type)
 - Post queries (sentry interface type)
 - Modified api for capture_exception

0.02 2014-05-23
 - Override event defaults in the constructor
 - Automatically post stack traces with the error handler
 - Post stack traces (sentry interface type)
 - Post http requests (sentry interface type)
 - Package improvements

0.01 2014-05-07
 - Initial release


lib/Sentry/Raven/Processor.pm
package Sentry::Raven::Processor;

use strict;

=head1 NAME

Sentry::Raven::Processor - Sentry event processors

=head1 SYNOPSIS

  use Sentry::Raven;
  use Sentry::Raven::Processor::RemoveStackVariables;

  my $raven = Sentry::Raven->new(
    processors => [ Sentry::Raven::Processor::RemoveStackVariables ],
  );

=head1 DESCRIPTION

Processors are a mechanism for modifying events after they are generated but before they are posted to the sentry service.  They are useful for scrubbing sensitive data, such as passwords, as well as adding additional context.

=cut

=head1 STANDARD PROCESSORS

=head1 IMPLEMENTING A PROCESSOR

Processors must have the following class methods:

=head2 $processed_event = My::Processor->process($event)

This is the interface for processing events.  Returns a modified version of the event.

=cut

1;


lib/Sentry/Raven/Processor/RemoveStackVariables.pm
package Sentry::Raven::Processor::RemoveStackVariables;

use 5.008;
use strict;
use warnings;

=head1 NAME

Sentry::Raven::Processor::RemoveStackVariables - Remove stack variables from stack traces in events

=head1 SYNOPSIS

  use Sentry::Raven;
  use Sentry::Raven::Processor::RemoveStackVariables;

  my $raven = Sentry::Raven->new(
    processors => [ Sentry::Raven::Processor::RemoveStackVariables ],
  );

=head1 DESCRIPTION

This processor removes variables from stack traces before they are posted to the sentry service.  This prevents sensitive values from being exposed, such as passwords or credit card numbers.

=head1 METHODS

=head2 my $processed_event = Sentry::Raven::Processor::RemoveStackVariables->process( $event )

Process an event.

=cut

sub process {
    my ($class, $event) = @_;
    if ($event->{'sentry.interfaces.Stacktrace'}) {
        my $num_frames = scalar(@{$event->{'sentry.interfaces.Stacktrace'}->{frames}});
        delete($event->{'sentry.interfaces.Stacktrace'}->{frames}->[$_]->{vars}) for 0..($num_frames-1);
    }
    return $event;
}

1;


Makefile.PL
use ExtUtils::MakeMaker;

WriteMakefile1(
    NAME                => 'Sentry::Raven',
    ABSTRACT_FROM       => 'lib/Sentry/Raven.pm',
    VERSION_FROM        => 'lib/Sentry/Raven.pm',
    MIN_PERL_VERSION    => '5.008',
    AUTHOR              => 'Matt Harrington',
    LICENSE             => 'mit',

    PREREQ_PM           => {
        'Data::Dump'                   => 0,
        'Devel::StackTrace'            => 0,
        'English'                      => 0,
        'File::Basename'               => 0,
        'File::Slurp'                  => 0,
        'HTTP::Request::Common'        => 0,
        'HTTP::Status'                 => 0,
        'JSON::XS'                     => 0,
        'List::Util'                   => 0,
        'LWP::Protocol::https'         => 0,
        'LWP::UserAgent'               => 0,
        'Moo'                          => 0,
        'MooX::Types::MooseLike::Base' => 0,
        'Sys::Hostname'                => 0,
        'Time::Piece'                  => 0,
        'URI'                          => 0,
        'UUID::Tiny'                   => 0,
    },

    TEST_REQUIRES       => {
        'File::Spec'                        => 0,
        'HTTP::Response'                    => 0,
        'IO::Uncompress::Gunzip'            => 0,
        'Test::CPAN::Changes::ReallyStrict' => 0,
        'Test::LWP::UserAgent'              => 0,
        'Test::More'                        => '0.88',
        'Test::Perl::Critic'                => '1.03',
        'Test::Warn'                        => '0.30',
    },

    META_MERGE          => {
        'meta-spec' => {
            version => '2',
            url     => 'http://search.cpan.org/perldoc?CPAN::Meta::Spec',
        },

        homepage    => 'https://github.com/getsentry/perl-raven',
        resources   => {
            repository => {
                type => 'git',
                url  => 'git@github.com:getsentry/perl-raven.git',
                web  => 'https://github.com/getsentry/perl-raven',
            },
        },
    },
);


sub WriteMakefile1 {   # Cribbed from eumm-upgrade by Alexandr Ciornii
  my %params = @_;
  my $eumm_version = $ExtUtils::MakeMaker::VERSION;
  $eumm_version = eval $eumm_version;

  if ($params{TEST_REQUIRES} and $eumm_version < 6.6303) {
      $params{BUILD_REQUIRES}={ %{$params{BUILD_REQUIRES} || {}} , %{$params{TEST_REQUIRES}} };
      delete $params{TEST_REQUIRES};
  }
  if ($params{BUILD_REQUIRES} and $eumm_version < 6.5503) {
      #EUMM 6.5502 has problems with BUILD_REQUIRES
      $params{PREREQ_PM}={ %{$params{PREREQ_PM} || {}} , %{$params{BUILD_REQUIRES}} };
      delete $params{BUILD_REQUIRES};
  }
  delete $params{CONFIGURE_REQUIRES} if $eumm_version < 6.52;
  delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
  delete $params{META_MERGE} if $eumm_version < 6.46;
  delete $params{META_ADD} if $eumm_version < 6.46;
  delete $params{LICENSE} if $eumm_version < 6.31;
  delete $params{AUTHOR} if $] < 5.005;
  delete $params{ABSTRACT_FROM} if $] < 5.005;
  delete $params{BINARY_LOCATION} if $] < 5.005;

  WriteMakefile(%params);
}
